src/hotspot/share/cds/heapShared.cpp

```diff
@@ -339,6 +353,28 @@ class KlassToOopHandleTable: public ResourceHashtable<Klass*, OopHandle,
   }
 };
 
+void HeapShared::add_scratch_resolved_reference(ConstantPool* src, int index, oop new_result) {
+  MutexLocker ml(ScratchObjects_lock, Mutex::_no_safepoint_check_flag);
+  OopHandle* scratch_handle = _scratch_references_table->get(src);
+  assert(scratch_handle != nullptr, "Must exist");
+  objArrayOop scratch = (objArrayOop)(scratch_handle->resolve());
+  scratch->obj_at_put(index, new_result);
+}
+
+void HeapShared::add_scratch_resolved_references(ConstantPool* src, OopHandle dest) {
+  MutexLocker ml(ScratchObjects_lock, Mutex::_no_safepoint_check_flag);
+  if (_scratch_references_table == nullptr) {
+    _scratch_references_table = new (mtClass)ResolvedReferenceScratchTable();
+  }
+  assert(_scratch_references_table != nullptr, "must be initialized");
+  _scratch_references_table->put(src, dest);
+}
+
+OopHandle HeapShared::scratch_resolved_references(ConstantPool* src) {
+  MutexLocker ml(ScratchObjects_lock, Mutex::_no_safepoint_check_flag);
+  return *(_scratch_references_table->get(src));
+}
+
```

Comment:
The logic of these two new functions are covered by the KlassToOopHandleTable class, so I would suggest refactoring   KlassToOopHandleTable to

```
class MetaspaceObjToOopHandleTable: public ResourceHashtable<MetaspaceObj*, OopHandle, ...
```

So your new "scratch_resolved_references" functions can be similar to `HeapShared::scratch_java_mirror/set_scratch_java_mirror`.

Also, this function needs to be updated

```
void HeapShared::remove_scratch_objects(Klass* k) {
  _scratch_java_mirror_table->remove_oop(k);
  if (k->is_instance_klass()) {
    _scratch_resolved_references_table->remove(InstanceKlass::cast(k)->constants());
  }
}
```